src/Microsoft.ML.Core/Data/ReadOnlyMemoryUtils.cs

```diff
@@ -0,0 +1,399 @@
+        {
+            if (memory.IsEmpty)
+                return memory;
+
+            int ichLim = memory.Length;
+            if (!char.IsWhiteSpace(memory.Span[ichLim - 1]))
+                return memory;
+
+            while (0 < ichLim && char.IsWhiteSpace(memory.Span[ichLim - 1]))
+                ichLim--;
+
+            return memory.Slice(0, ichLim);
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(out Single value, ReadOnlyMemory<char> memory)
+        {
+            var res = DoubleParser.Parse(out value, memory);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        /// <summary>
+        /// This produces zero for an empty string.
+        /// </summary>
+        public static bool TryParse(out Double value, ReadOnlyMemory<char> memory)
+        {
+            var res = DoubleParser.Parse(out value, memory);
+            Contracts.Assert(res != DoubleParser.Result.Empty || value == 0);
+            return res <= DoubleParser.Result.Empty;
+        }
+
+        public static uint Hash(uint seed, ReadOnlyMemory<char> memory) => Hashing.MurmurHash(seed, memory);
+
+        public static NormStr AddToPool(NormStr.Pool pool, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            return pool.Add(memory);
+        }
+
+        public static NormStr FindInPool(NormStr.Pool pool, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(pool, nameof(pool));
+            return pool.Get(memory);
+        }
+
+        public static void AddToStringBuilder(StringBuilder sb, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(sb, nameof(sb));
+            if (!memory.IsEmpty)
+                sb.AppendAll(memory);
+        }
+
+        public static void AddLowerCaseToStringBuilder(StringBuilder sb, ReadOnlyMemory<char> memory)
+        {
+            Contracts.CheckValue(sb, nameof(sb));
+
+            if (!memory.IsEmpty)
+            {
+                int ichLim = memory.Length;
+                int min = 0;
+                int j;
+                for (j = min; j < ichLim; j++)
+                {
+                    char ch = CharUtils.ToLowerInvariant(memory.Span[j]);
+                    if (ch != memory.Span[j])
+                    {
+                        sb.Append(memory, min, j - min).Append(ch);
+                        min = j + 1;
+                    }
+                }
+
+                Contracts.Assert(j == ichLim);
+                if (min != j)
+                    sb.Append(memory, min, j - min);
+            }
+        }
+
+        // REVIEW: Can this be faster?
+        private static bool ContainsChar(char ch, char[] rgch)
+        {
+            Contracts.CheckNonEmpty(rgch, nameof(rgch));
```

Comment:
>Contracts.CheckNonEmpty(rgch, nameof(rgch)); [](start = 12, length = 44)

Since there's that review comment (though I suspect it was just copy/pasted over), yes, this can be faster.

First this is a private method that you're calling repeatedly in tight loops in the split code. An *assert* on the non-emptyness of the input is appropriate for sure, and you ought to check that the code is unreachable in any situation where the separators fed in should not be null, but other than that it seems fine. A check is code run all the time even in a release build, which is not needed here.